Tests/FluentTests/QueryHistoryTests.swift

```diff
@@ -0,0 +1,153 @@
+        defer { app.shutdown() }
+
+        let test = ArrayTestDatabase()
+        app.databases.use(test.configuration, as: .test)
+
+        test.append([
+            TestOutput(["id": 1, "content": "a"]),
+            TestOutput(["id": 2, "content": "b"]),
+        ])
+
+        app.get("foo") { req -> EventLoopFuture<[Post]> in
+            return Post.query(on: req.db).all().map { posts in
+                XCTAssertEqual(req.fluent.history.queries.count, 0)
+                return posts
+            }
+        }
+
+        try app.testable().test(.GET, "foo") { res in
+            XCTAssertEqual(res.status, .ok)
+        }
+    }
+
+    func testQueryHistoryEnabled() throws {
+        let app = Application(.testing)
+        defer { app.shutdown() }
+
+        let test = ArrayTestDatabase()
+        app.databases.use(test.configuration, as: .test)
+
+        test.append([
+            TestOutput(["id": 1, "content": "a"]),
+            TestOutput(["id": 2, "content": "b"]),
+        ])
+
+        app.get("foo") { req -> EventLoopFuture<[Post]> in
+            req.fluent.history.start()
+            return Post.query(on: req.db).all().map { posts in
+                XCTAssertEqual(req.fluent.history.queries.count, 1)
+                return posts
+            }
+        }
+
+        try app.testable().test(.GET, "foo") { res in
+            XCTAssertEqual(res.status, .ok)
+        }
+    }
+
+    func testQueryHistoryEnableAndDisable() throws {
+        let app = Application(.testing)
+        defer { app.shutdown() }
+
+        let test = ArrayTestDatabase()
+        app.databases.use(test.configuration, as: .test)
+
+        test.append([
+            TestOutput(["id": 1, "content": "a"]),
+            TestOutput(["id": 2, "content": "b"]),
+        ])
+
+        app.get("foo") { req -> EventLoopFuture<[Post]> in
+            req.fluent.history.start()
+            return Post.query(on: req.db).all().flatMap { posts -> EventLoopFuture<[Post]> in
+                XCTAssertEqual(req.fluent.history.queries.count, 1)
+                req.fluent.history.stop()
+
+                test.append([
+                    TestOutput(["id": 1, "content": "a"]),
+                    TestOutput(["id": 2, "content": "b"]),
+                ])
+
+                return Post.query(on: req.db).all()
+            }.map { posts in
+                XCTAssertEqual(req.fluent.history.queries.count, 0)
+                return posts
+            }
+        }
+
+        try app.testable().test(.GET, "foo") { res in
+            XCTAssertEqual(res.status, .ok)
+        }
+    }
+
+    func testQueryHistoryForApp() throws {
+        let app = Application(.testing)
+        defer { app.shutdown() }
+
+        app.fluent.history.start()
+        let test = ArrayTestDatabase()
+        app.databases.use(test.configuration, as: .test)
+
+        app.get("foo") { req -> EventLoopFuture<[Post]> in
+            return Post.query(on: app.db).all()
```

Comment:
You can just do `Post.query(on: app.db).all().wait()` instead of calling `app.test`. That way we're not using `app.db` inside of a route handler which could cause problems.